File: src/com/company/tree/isBST.java
package com.company.tree;

public class isBST {

    public boolean isValidBST(TreeNode root) {
        return isValidBSTUtils(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }


    private boolean isValidBSTUtils(TreeNode root, long minValue, long maxValue) {
        if (root == null)
            return true;
        return (long) root.value > minValue && (long) root.value < maxValue
                && isValidBSTUtils(root.left, minValue, root.value)
                && isValidBSTUtils(root.right, root.value, maxValue);
    }
}

File: src/com/company/tree/isBST.kt
package com.company.tree

private fun isValidBSTUtils(root: TreeNode?, minValue: Long, maxValue: Long): Boolean {
    return if (root == null) {
        true
    } else {
        root.`val`.toLong() > minValue && root.`val`.toLong() < maxValue &&
                isValidBSTUtils(root.left, minValue, root.`val`.toLong()) &&
                isValidBSTUtils(root.right, root.`val`.toLong(), maxValue)
    }
}

fun isValidBST(root: TreeNode?): Boolean {
    return isValidBSTUtils(root, Long.MIN_VALUE, Long.MAX_VALUE)
}
